### PR TITLE
#478 Update link for shared libraries

### PR DESCRIPTION
**Pull request by lecourtoisflo**
I don't know how to test locally if it resolves the issue of trace singleton on Windows.

Impact on libraries that are shared in all cases (solvers + models)

### CODE REVIEW

**Comment by gautierbureau** [comment]
For me it does not solve the issue. The singletons have to be put outside of dynawo_Common in a new library dynawo_Util (corresponding to the log lib in my poc)

**Comment by lecourtoisflo** [comment]
> For me it does not solve the issue. The singletons have to be put outside of dynawo_Common in a new library dynawo_Util (corresponding to the log lib in my poc)



> For me it does not solve the issue. The singletons have to be put outside of dynawo_Common in a new library dynawo_Util (corresponding to the log lib in my poc)

I don't get the point of creating a new lib here. What is the difference ? I don't think we should gather all singletons in a lib for technical reasons like these, this will make the code organization less clear for me

**Comment by gautierbureau** [comment]
> > For me it does not solve the issue. The singletons have to be put outside of dynawo_Common in a new library dynawo_Util (corresponding to the log lib in my poc)
> 
> > For me it does not solve the issue. The singletons have to be put outside of dynawo_Common in a new library dynawo_Util (corresponding to the log lib in my poc)
> 
> I don't get the point of creating a new lib here. What is the difference ? I don't think we should gather all singletons in a lib for technical reasons like these, this will make the code organization less clear for me

The difference is everything... The symbols of singletons should not be present in a lib that models are linked to or we need dynawo_Common for models so the singletons need to be out of it. The dynawo_Util was already there before refactoring from Frederic which was a mistake it really needs to exist. The issue is a technical issue that needs to be solved by a technical solution.

**Comment by lecourtoisflo** [comment]
> > > For me it does not solve the issue. The singletons have to be put outside of dynawo_Common in a new library dynawo_Util (corresponding to the log lib in my poc)
> > 
> > 
> > > For me it does not solve the issue. The singletons have to be put outside of dynawo_Common in a new library dynawo_Util (corresponding to the log lib in my poc)
> > 
> > 
> > I don't get the point of creating a new lib here. What is the difference ? I don't think we should gather all singletons in a lib for technical reasons like these, this will make the code organization less clear for me
> 
> The difference is everything... The symbols of singletons should not be present in a lib that models are linked to and we need dynawo_Common for models so the singletons need to be out of it. The dynawo_Util was already there before refactoring from Frederic which was a mistake it really needs to exist. The issue is a technical issue that needs to be solved by a technical solution.

models are not linked to dynawo_Common any more. That's precisely the modification

File: dynawo/sources/Models/CMakeLists.txt [hunk 1]
```diff
@@ -294,12 +294,21 @@ macro(COMPILE_CPP_MODEL
...
     target_link_libraries(${CPP_MODEL}
       PRIVATE
-        dynawo_Common
         dynawo_ModelerCommon
       PUBLIC
         dynawo_CPPModelsCommon
       )
 
+    if(LIBRARY_TYPE STREQUAL "STATIC")
+      target_include_directories(lib PRIVATE $<TARGET_PROPERTY:dynawo_Common,INTERFACE_INCLUDE_DIRECTORIES>)
```
Comment (gautierbureau) [inline dynawo/sources/Models/CMakeLists.txt:303]: lib is a wrong copy/paste

File: dynawo/sources/Models/CMakeLists.txt [hunk 2]
```diff
@@ -294,12 +294,21 @@ macro(COMPILE_CPP_MODEL
...
+      if(MSVC OR APPLE)
+        # Link done to avoid duplicating the Trace singleton
+        target_link_libraries(${CPP_MODEL} PRIVATE dynawo_Common)
+      endif()
+    else()
+      target_link_libraries(${CPP_MODEL} PRIVATE dynawo_Common)
```
Comment (gautierbureau) [inline dynawo/sources/Models/CMakeLists.txt:309]: you are still linking to dynawo_Common here

**Review by gautierbureau — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/CMakeLists.txt (hunk 1 above)
Comment (lecourtoisflo) [reply dynawo/sources/Models/CMakeLists.txt:303]: done

File: dynawo/sources/Models/CMakeLists.txt (hunk 2 above)
Comment (lecourtoisflo) [reply dynawo/sources/Models/CMakeLists.txt:309]: Correct, there is an error. But in that case, I don't get to which target it is supposed to link (an executable ? another library ?)

Comment (gautierbureau) [reply dynawo/sources/Models/CMakeLists.txt:309]: In shared you can link to the library containing the singleton (here that would still be dynawo_Common or dynawo_Util as I propose) and in static you need to link to the executable

**Comment by lecourtoisflo** [comment]
not releveant for now. It is easier to re-create it later

**Pull request closed without merging**